fdbclient/ManagementAPI.actor.cpp

```diff
@@ -1587,6 +1589,69 @@ ACTOR Future<Void> excludeServers(Database cx, vector<AddressExclusion> servers,
 				wait(tr.commit());
 				return Void();
 			} catch (Error& e) {
+				TraceEvent("ExcludeServersError").error(e, true);
+				wait(tr.onError(e));
+			}
+		}
+	}
+}
+
+void excludeLocalities(Transaction& tr, std::unordered_set<std::string>& localities, bool failed) {
+	tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+	tr.setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+	tr.setOption(FDBTransactionOptions::LOCK_AWARE);
+	tr.setOption(FDBTransactionOptions::USE_PROVISIONAL_PROXIES);
+	std::string excludeVersionKey = deterministicRandom()->randomUniqueID().toString();
+	auto localityVersionKey = failed ? failedLocalityVersionKey : excludedLocalityVersionKey;
+	tr.addReadConflictRange(singleKeyRange(localityVersionKey)); // To conflict with parallel includeLocalities
+	tr.set(localityVersionKey, excludeVersionKey);
+	for (auto& l : localities) {
+		if (failed) {
+			tr.set(encodeFailedLocalityKey(l), StringRef());
+		} else {
+			tr.set(encodeExcludedLocalityKey(l), StringRef());
+		}
+	}
+	TraceEvent("ExcludeLocalitiesCommit")
+		.detail("Localities", describe(localities))
+		.detail("ExcludeFailed", failed);
+}
+
+ACTOR Future<Void> excludeLocalities(Database cx, std::unordered_set<std::string> localities, bool failed) {
```

Comment:
It could alternatively be passed by pointer, but you can't pass it by reference because the actor compiler implicitly converts arguments to being passed by reference and then copies them, I believe.